servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -146,27 +155,37 @@
         resolver = builder.build();
     }
 
+    @Nullable
+    private DnsDiscoveryObserver newDiscoveryObserver(final String address) {
+        return observer != null ? observer.newDiscovery(address) : null;
+    }
+
     @Override
     public Publisher<ServiceDiscovererEvent<InetAddress>> dnsQuery(final String address) {
-        return new ARecordPublisher(true, address).flatMapConcatIterable(identity());
+        requireNonNull(address);
+        return defer(() -> new ARecordPublisher(true, address, newDiscoveryObserver(address))
```

Comment:
This will create a new `DnsDiscoveryObserver` every time a retry strategy resubscribes. This is not ideal, but will work in consistent way with `dnsSrvQuery`.
Alternative approaches I see:
1. Create a `DnsDiscoveryObserver` when users invoke `ServiceDiscoverer#discover(...)`. Because `ARecordPublisher` does not support multiple subscribers anyway, that will work fine. However, the behavior will be different from `dnsSrvQuery`.
2. Create a `DnsDiscoveryObserver` inside `handleSubscribe0`. It will require rework of `dnsSrvQuery` to not trigger `newResolution` for subsequent A queries.

Reply:
For now, I trigger `newResolution` for subsequent A queries of SRV discoverer to keep the scope of this PR minimal and avoid changing the SRV discoverer code.
Since SRV is not the most popular use-case it should be ok to defer enhancements there.

Reply:
Since we can not distinguish between a retry vs a new subscribe, so creating a new observer per subscribe makes sense to me.
Since, users control the scope of `DnsServiceDiscovererObserver` and association with a client, if the same observer instance needs to be shared across retries/subscribes, they can do so by returning the same observer instance from `newDiscovery()`?

Reply:
>  if the same observer instance needs to be shared across retries/subscribes, they can do so by returning the same observer instance from `newDiscovery()`?

Correct, this is the way for users if necessary.